R/tool-read-web.R

```diff
@@ -0,0 +1,220 @@
+#' @include tool-result.R
+NULL
+
+#' Tool: Read a Web Page as Markdown
+#'
+#' @examplesIf rlang::is_installed("chromote") && rlang::is_interactive()
+#' btw_tool_web_read_url("https://www.r-project.org/")
+#' btw_tool_web_read_url(
+#'   "https://posit.co/blog/easy-tool-calls-with-ellmer-and-chatlas/"
+#' )
+#'
+#' @param url The URL of the web page to read.
+#' @param ... Ignored, for future features.
+#' @param max_wait_for_page_load_s Maximum time to wait for the page to load, in
+#'   seconds. Can be set globally using the `btw.max_wait_for_page_load_s`
+#'   option.
+#'
+#' @family Tools
+#' @export
+btw_tool_web_read_url <- function(
+  url,
+  ...,
+  max_wait_for_page_load_s = getOption("btw.max_wait_for_page_load_s", 10)
+) {
+  html <- read_url_main_content(url, timeout = max_wait_for_page_load_s)
+
+  if (is.null(html) || !nzchar(html)) {
+    cli::cli_abort(
+      "Failed to read web page at {.url {url}}. Please check the URL."
+    )
+  }
+
+  md <- paste(pandoc_html_simplify(html), collapse = "\n")
+  res <- glue_('<web_page_content url="{{url}}">\n{{md}}\n</web_page_content>')
+
+  BtwWebPageResult(res)
+}
+
+BtwWebPageResult <- S7::new_class(
+  "BtwWebPageResult",
+  parent = BtwToolResult
+)
+
+has_chromote <- function() {
+  tryCatch(
+    {
+      rlang::check_installed("chromote", version = "0.5.1.9000")
+      TRUE
+    },
+    error = function(e) {
+      cli::cli_warn(
+        "The `chromote` package is required to use the web reading tool."
+      )
+      FALSE
+    }
+  )
+}
+
+.btw_add_to_tools(
+  name = "btw_tool_web_read_url",
+  group = "web",
+  tool = function() {
+    if (!has_chromote()) {
+      return(NULL)
+    }
+
+    ellmer::tool(
+      btw_tool_web_read_url,
+      .description = 'Read a web page and convert it to Markdown format.
+
+This tool fetches the content of a web page and returns it as a simplified Markdown representation.
+
+WHEN TO USE: Use this tool when you need to access and analyze the content of a web page, e.g. when the user asks you to read the contents of a webpage.',
+      .annotations = ellmer::tool_annotations(
+        title = "Read Web Page",
+        read_only_hint = TRUE,
+        open_world_hint = TRUE,
+        idempotent_hint = FALSE
+      ),
+      url = ellmer::type_string(
+        "The URL of the web page to read."
+      )
+    )
+  }
+)
+
+read_url_main_content <- function(url, timeout = 10) {
+  b <- chromote::ChromoteSession$new()
```

Comment:
What are `b` and `p` shorthand for? Could we use a slightly longer variable name there?

Reply:
It's unfortunately common convention when working with chromote. `b` stands for browser and `p` for a promise. I sometimes name these things differently but often find it adds more cognitive overhead because so much other chromote code follows this convention.

Reply:
Got it!